Make ToolConfWatcher watch ToolCache

mvdbeek opened the pull request:
If any item in the tool cache has changed (local tool update, repository
update, tool deleted) the toolbox will be reloaded and any tool changes are
applied.

Should fix #3813, ping @blankenberg 

I have tested this with both data managers and regular repository updates. You can also test this by modifying any loaded tool, these will automatically be updated now (Is there any reason left to keep the reload toolbox / reload tool xml file buttons around?).

mvdbeek committed:
Make ToolConfWatcher watch ToolCache

If any item in the tool cache has changed (local tool update, repository
update, tool deleted) the toolbox will be reloaded and any tool changes are
applied.

Should fix https://github.com/galaxyproject/galaxy/issues/3813.

mvdbeek committed:
Fix ToolConfWatcher test: can only watch existing files

mvdbeek committed:
Only call tool_cache.cleanup if tool_cache exists

mvdbeek committed:
Fix toolshed startup by setting tool_cache to `None` (the toolshed loads an empty toolbox)

mvdbeek committed:
Wrap tool_cache cleanup in try/except clause

nsoranzo commented on lib/galaxy/tools/toolbox/cache.py:
```diff
@@ -35,10 +59,12 @@ def expire_tool(self, tool_id):
         tool_hash = md5_hash_file(config_filename)
         tool_id = str( tool.id )
         self._hash_by_tool_paths[config_filename] = tool_hash
+        self._mod_time_by_path[config_filename] = time.ctime(os.path.getmtime(config_filename))
```
Maybe it's better to leave it as seconds since epoch instead of local time? Local time would make sense only if we expose this value to a human. Also, if it is epoch time you can change the comparison in line 46 to use `<` .

nsoranzo commented on lib/galaxy/tools/toolbox/cache.py:
```diff
@@ -13,16 +14,39 @@ def __init__(self):
+            return []
+
+    def _should_cleanup(self, config_filename):
+        """Return True of `config_filename` does not exist or if modtime and hash have changes, else return False."""
```
s/of/if/

nsoranzo commented on lib/galaxy/tools/toolbox/cache.py:
```diff
@@ -13,16 +14,39 @@ def __init__(self):
+        if not os.path.exists(config_filename):
+            return True
+        new_mtime = time.ctime(os.path.getmtime(config_filename))
+        if self._mod_time_by_path.get(config_filename) != new_mtime:
```
Did you add the check on mtime to make the method faster? Is `md5_hash_file()` slow even on the usually small tool XML files?

mvdbeek replied:
I was assuming it would be slightly faster to look at the mtime than to read in all tool xml files and calculate the hash on them. Though I think this takes a negligible amount of time in both cases. We do the same thing in the ToolConfWatcher (it was initially only mtime, I added the hashing to not trigger unnecessary reloads during the install process).

nsoranzo commented on lib/galaxy/tools/toolbox/cache.py:
```diff
@@ -13,16 +14,39 @@ def __init__(self):
+        """
+        try:
+            paths_to_cleanup = {path: tool.all_ids for path, tool in self._tools_by_path.items() if self._should_cleanup(path)}
+            removed_tool_ids = []
```
Move this line before the `try:`

nsoranzo commented on lib/galaxy/tools/toolbox/cache.py:
```diff
@@ -13,16 +14,39 @@ def __init__(self):
+            paths_to_cleanup = {path: tool.all_ids for path, tool in self._tools_by_path.items() if self._should_cleanup(path)}
+            removed_tool_ids = []
+            for config_filename, tool_ids in paths_to_cleanup.items():
+                removed_tool_ids.extend(tool_ids)
```
Move this line after the next `for` loop.

nsoranzo commented on lib/galaxy/tools/toolbox/cache.py:
```diff
@@ -13,16 +14,39 @@ def __init__(self):
+        except Exception:
+            # If by chance the file is being removed while calculating the hash or modtime
+            # we don't want the thread to die.
+            return []
```
So you can `return removed_tool_ids`

nsoranzo commented on lib/galaxy/tools/toolbox/base.py:
```diff
@@ -79,7 +79,13 @@ def __init__( self, config_filenames, tool_root_dir, app, tool_conf_watcher=None
+                tool_cache = self.app.tool_cache
+            else:
+                tool_cache = None
+            self._tool_conf_watcher = get_tool_conf_watcher(reload_callback=lambda: self.handle_reload_toolbox(), tool_cache=tool_cache)
```
You may also want to modify the `get_tool_conf_watcher()` calls in `lib/galaxy/tools/data_manager/manager.py`

mvdbeek replied:
It's enough to do it once for the toolbox, since the data managers are loaded in the tool cache like regular tools. I prefer not checking the tool_cache multiple times, but if you think it's better for consistency I can do that.

nsoranzo replied:
I get it now, we may think about a separate tool_cache watcher.

mvdbeek replied:
That was my first instinct, but I was thinking that it may be better to have just a single thread watching for toolbox reloads. If we trigger a toolbox reload because the shed_tool_conf and the tool xml have changed, we reload just once now, but if we have separate threads doing the same thing we would need to synchronize them somehow.

nsoranzo replied:
Actually after posting the previous comment I thought that maybe it would be better to have a single toolbox watcher monitoring tool and data manager config files plus the tool_cache, i.e. get rid of the data manager separate tool conf watcher. It seems we are on the same line!

mvdbeek replied:
Yep, I'll have a look at doing that.

mvdbeek committed:
Improve ToolCache cleanup logic

Pull `removed_tool_ids` out of try/except, so that we can return
`removed_tool_ids` even if an exception occured. Fix a typo in the
`_should_cleanup` docstring. Many thanks for the suggestions @nsoranzo.

blankenberg commented:
Awesome, thanks @mvdbeek, and @nsoranzo for the review!
